lib/llvmopencl/EmitPrintf.cc

```diff
@@ -0,0 +1,679 @@
+//===- AMDGPUEmitPrintf.cpp -----------------------------------------------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// Utility function to lower a printf call into a series of device
+// library calls on the AMDGPU target.
+//
+// WARNING: This file knows about certain library functions. It recognizes them
+// by name, and hardwires knowledge of their semantics.
+//
```

Comment:
Please add a note that this is copied from LLVM and should be retained in sync over the versions with upstreaming planned. We likely have to exclude it from clang-format too if the LLVM formatting doesn't much ours. In fact we should fix our clang-format script in that case.

Reply:
We should keep it nicely diffable to AMDGPUEmitPrintf.cpp with the diff minimized over time.